include/LightGBM/utils/common.h

```diff
@@ -1082,12 +1082,18 @@ struct __StringToTHelper<T, true> {
     // Fast (common) path: For numeric inputs in RFC 7159 format:
     const bool fast_parse_succeeded = fast_double_parser::parse_number(str.c_str(), &tmp);
 
-    // Rare path: Not in RFC 7159 format. Possible "inf", "nan", etc. Fallback to standard library:
+    // Rare path: Not in RFC 7159 format. Possible "inf", "nan", etc.
     if (!fast_parse_succeeded) {
-      std::stringstream ss;
-      Common::C_stringstream(ss);
-      ss << str;
-      ss >> tmp;
+      std::string strlower(str); 
+      std::transform(strlower.begin(), strlower.end(), strlower.begin(), [](int c) -> char { return static_cast<char>(::tolower(c)); });
```

Comment:
Great clean code @mjmckp ;)

Instead of allocating a string, and since you already have a lambda, what about defining a lower-case comparison lambda and use `std::equal` to check the "inf" and "nan" values below?

Although this is the rare branch there might be longer strings than inf or nan which might be parsed here and might slow down our parsing without need.

Reply:
Example: https://thispointer.com/c-case-insensitive-string-comparison-using-stl-c11-boost-library/

Reply:
Again, this hardly seems worth it, this branch is rarely invoked, meanwhile a colossal amount of strings are being allocated in splitting and parsing the input file, so these few extra allocations are a drop in the ocean.

I think our time is better spent adding robust round-trip tests to ensure major bugs like this don't occur again...

Reply:
> I think our time is better spent adding robust round-trip tests to ensure major bugs like this don't occur again...

Agreed. Will you add such tests?
I actually run such tests but on an external lgbm provider and didn't have nan nor inf on my model, but would prefer to see them in lgbm's CI so any breakage is detected immediately.

Reply:
I could assist in adding the tests, any idea where these should go and how best to implement them?

Reply:
@mjmckp Thank you! For tests we need to merge #3555 first, I guess.

Reply:
@mjmckp #3555 has been just merged. We will really appreciate new GTest-compatible tests in this or in a follow-up PR.

Reply:
Sure, would you mind pointing me towards a similar kind of test that I can use as a starting point please?

Reply:
Unfortunately, we don't have any tests yet. This is something that we should concentrate on in the near future. For now, I think you can take a look at tests from @AlbertoEAF in #3997. 
https://github.com/microsoft/LightGBM/pull/3997/files#diff-c363eba6eda99d9e560f8341a1fc8fe02e885d2256db2482e1c543430a25666d

Reply:
@mjmckp I've merged this PR with the aim to not delay the upcoming release. Please feel free to add tests in a new PR. We'll be very grateful! And thanks a lot for the bug fix!

Reply:
@StrikerRUS Thanks a lot, I'll get up to speed on how the new tests work and add some tests for this in a new PR soon.